File: modern/src/main/rs/emulate/modern/script/interp/InstructionMap.java
package rs.emulate.modern.script.interp;

import com.google.common.collect.ImmutableMap;
import rs.emulate.modern.script.decomp.instr.Instruction;
import rs.emulate.modern.script.decomp.instr.impl.BranchInstruction;
import rs.emulate.modern.script.decomp.instr.impl.GenericInstruction;
import rs.emulate.modern.script.decomp.instr.impl.IntInstruction;
import rs.emulate.modern.script.decomp.instr.impl.StringInstruction;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/**
 * A Map of opcodes to {@link Instruction}s.
 *
 * @author Major
 */
public final class InstructionMap {

	/**
	 * A creator for a {@link Map} of instruction opcodes to {@link Consumer}s, which modify a {@link ScriptContext}
	 * using side-effects.
	 */
	private static final class InstructionMapCreator {

		/**
		 * The Map of opcodes to ScriptContext Consumers.
		 */
		private final Map<Integer, Instruction> instructions = new HashMap<>(100);

		/**
		 * Creates the {@link Map} of opcodes to {@link Consumer}s.
		 *
		 * @return The Map.
		 */
		public Map<Integer, Instruction> create() {
			return ImmutableMap.copyOf(instructions);
		}

		/**
		 * Fills the instruction Map.
		 */
		public void fill() {
			insert(IntInstruction.create("pushi", 0, context -> context.pushInt(context.getIntOperand())));
			insert(StringInstruction.create("pushs", 3, context -> context.pushString(context.getStringOperand())));

			insert(BranchInstruction.create("goto", 6, ScriptContext::branch));
			insert(BranchInstruction.create("ifi_neq", 7, context -> context.branchIf(context.popInt() != context.popInt())));
			insert(BranchInstruction.create("ifi_eq", 8, context -> context.branchIf(context.popInt() == context.popInt())));
			insert(BranchInstruction.create("ifi_lt", 9, context -> context.branchIf(context.popInt() > context.popInt())));
			insert(BranchInstruction.create("ifi_gt", 10, context -> context.branchIf(context.popInt() < context.popInt())));

			insert(BranchInstruction.create("ifi_geq", 31, context -> context.branchIf(context.popInt() <= context.popInt())));
			insert(BranchInstruction.create("ifi_leq", 32, context -> context.branchIf(context.popInt() >= context.popInt())));

			insert(IntInstruction.create("concat", 37, context -> context.getStrings(context.stringCount() - context.popInt())
					.collect(Collectors.joining())));

			insert(GenericInstruction.create("popi", 38, ScriptContext::popInt));
			insert(GenericInstruction.create("pops", 39, ScriptContext::popString));

			insert(IntInstruction.create("pushl", 54, context -> context.pushLong(context.getLongOperand())));
			insert(IntInstruction.create("popl", 55, ScriptContext::popLong));

			insert(BranchInstruction.create("ifl_neq", 68, context -> context.branchIf(context.popLong() != context.popLong())));
			insert(BranchInstruction.create("ifl_eq", 69, context -> context.branchIf(context.popLong() == context.popLong())));
			insert(BranchInstruction.create("ifl_lt", 70, context -> context.branchIf(context.popLong() > context.popLong())));
			insert(BranchInstruction.create("ifl_gt", 71, context -> context.branchIf(context.popLong() < context.popLong())));

			insert(BranchInstruction.create("ifl_geq", 72, context -> context.branchIf(context.popLong() <= context.popLong())));
			insert(BranchInstruction.create("ifl_leq", 73, context -> context.branchIf(context.popLong() >= context.popLong())));

			insert(BranchInstruction.create("if_true", 86, context -> context.branchIf(context.popInt() == 1)));
			insert(BranchInstruction.create("if_false", 87, context -> context.branchIf(context.popInt() == 0)));
		}

		/**
		 * Inserts the specified {@link Instruction} into the {@link Map}.
		 *
		 * @param instruction The Instruction.
		 */
		private void insert(Instruction instruction) {
			instructions.put(instruction.getOpcode(), instruction);
		}

	}

	/**
	 * The map of opcodes to Consumers which perform functionality.
	 */
	private final Map<Integer, Instruction> instructions;

	/**
	 * Creates the InstructionMap.
	 */
	public InstructionMap() {
		InstructionMapCreator creator = new InstructionMapCreator();
		creator.fill();
		instructions = creator.create();
	}

	/**
	 * Executes the specified opcode for the specified {@link ScriptContext}.
	 *
	 * @param opcode The opcode.
	 * @param script The script.
	 */
	public void execute(int opcode, ScriptContext script) {
		Instruction function = instructions.get(opcode);
		function.evaluate(script);
	}

}

File: modern/src/main/rs/emulate/modern/script/interp/InstructionMap.kt
package rs.emulate.modern.script.interp

import com.google.common.collect.ImmutableMap
import rs.emulate.modern.script.decomp.instr.Instruction
import rs.emulate.modern.script.decomp.instr.OperandType
import rs.emulate.modern.script.decomp.instr.impl.BranchInstruction
import rs.emulate.modern.script.decomp.instr.impl.GenericInstruction
import rs.emulate.modern.script.decomp.instr.impl.IntInstruction
import rs.emulate.modern.script.decomp.instr.impl.LongInstruction
import rs.emulate.modern.script.decomp.instr.impl.StringInstruction
import java.util.HashMap
import java.util.stream.Collectors

/**
 * A Map of opcodes to [Instruction]s.
 */
class InstructionMap {

    /**
     * The map of opcodes to functionality.
     */
    private val instructions: Map<Int, Instruction>

    /**
     * A creator for a [Map] of instruction opcodes to functions, that modify a [ScriptContext] using side-effects.
     */
    private class InstructionMapCreator {

        private val instructions = HashMap<Int, Instruction>(100)

        fun create(): Map<Int, Instruction> {
            return ImmutableMap.copyOf(instructions)
        }

        /**
         * Fills the instruction Map.
         */
        fun fill() {
            int("pushi", 0) { pushInt(intOperand) }
            string("pushs", 3) { pushString(stringOperand) }

            branch("goto", 6) { branch() }
            branch("ifi_neq", 7) { branchIf(popInt() != popInt()) }
            branch("ifi_eq", 8) { branchIf(popInt() == popInt()) }
            branch("ifi_lt", 9) { branchIf(popInt() > popInt()) }
            branch("ifi_gt", 10) { branchIf(popInt() < popInt()) }

            branch("ifi_geq", 31) { branchIf(popInt() <= popInt()) }
            branch("ifi_leq", 32) { branchIf(popInt() >= popInt()) }

            int("concat", 37) { getStrings(stringCount() - popInt()).collect(Collectors.joining()) }

            generic("popi", 38) { popInt() }
            generic("pops", 39) { popString() }

            int("pushl", 54) { pushLong(longOperand) }
            int("popl", 55) { popLong() }

            branch("ifl_neq", 68) { branchIf(popLong() != popLong()) }
            branch("ifl_eq", 69) { branchIf(popLong() == popLong()) }
            branch("ifl_lt", 70) { branchIf(popLong() > popLong()) }
            branch("ifl_gt", 71) { branchIf(popLong() < popLong()) }

            branch("ifl_geq", 72) { branchIf(popLong() <= popLong()) }
            branch("ifl_leq", 73) { branchIf(popLong() >= popLong()) }

            branch("if_true", 86) { branchIf(popInt() == 1) }
            branch("if_false", 87) { branchIf(popInt() == 0) }
        }

        private fun branch(name: String, opcode: Int, action: ScriptContext.() -> Unit) {
            insert(BranchInstruction(name, opcode, action))
        }

        private fun generic(name: String, opcode: Int, vararg types: OperandType, action: ScriptContext.() -> Unit) {
            insert(GenericInstruction(name, opcode, action, *types))
        }

        private fun int(name: String, opcode: Int, action: ScriptContext.() -> Unit) {
            insert(IntInstruction(name, opcode, action))
        }

        private fun long(name: String, opcode: Int, action: ScriptContext.() -> Unit) {
            insert(LongInstruction(name, opcode, action))
        }

        private fun string(name: String, opcode: Int, action: ScriptContext.() -> Unit) {
            insert(StringInstruction(name, opcode, action))
        }

        private fun insert(instruction: Instruction) {
            instructions[instruction.opcode] = instruction
        }

    }

    /**
     * Creates the InstructionMap.
     */
    init {
        val creator = InstructionMapCreator()
        creator.fill()
        instructions = creator.create()
    }

    /**
     * Executes the specified opcode for the specified [ScriptContext].
     */
    fun execute(opcode: Int, script: ScriptContext) {
        val function = instructions[opcode] ?: throw IllegalStateException("Unrecognised instruction $opcode.")
        function.evaluate(script)
    }

}
